### ISSUE TITLE
Jar cannot be uninstalled.

### ISSUE THREAD

**Issue by lygoing** [jboss-javassist/javassist#213]
Chibash Hello, thank you very much for your work; I have a headache in a few days, my application needs to be hot deployed, need to uninstall the jar package during the run, but after using javassist (javassist loaded and modified I want Uninstalled jar package), so my application can not be deployed hot! Is there any solution? I really can't do anything. I look forward to your reply, thank you.

**Comment by chibash** [comment]
According to my quick inspection, all jar files seem to be properly closed.
How did you modify a jar file by Javassist?  Did you close it after the modification?

**Comment by lygoing** [comment]
@chibash Thank you very much for your answer.

Because the project code is more complicated, I just wrote a simple example I used.
[](https://github.com/lygoing/javassist-demo)

The important code is as follows:
`
public static void main(String[] args) throws Exception {
        String className = "com.zaxxer.hikari.HikariDataSource";

        CtClass ctClass = ClassPool.getDefault().getCtClass(className);
        CtConstructor ctConstructor = ctClass.makeClassInitializer();
        ctConstructor.setBody("System.out.println(\"Hello Javassist!\");");

        byte[] bytes = ctClass.toBytecode();
        ctClass.detach();

        //Write new class file
        File jarFile = getJarFile(className);
        String newClassPath = jarFile.getParentFile().getAbsolutePath() +
                File.separator + className.replace('.', File.separatorChar) + ".class";
        FileUtils.writeByteArrayToFile(new File(newClassPath), bytes);

        CommandLine commandLine = new CommandLine();

        //Uninstall the application jar package
        URLClassLoader classLoader = (URLClassLoader) JavassistExamples.class.getClassLoader();
        classLoader.close();

        //Write new class file to the jar package
        String[] command = new String[]{"jar", "-uvf", jarFile.getName(), className.replace('.', '/') + ".class"};
        String[] output = commandLine.execute(Runtime.getRuntime().exec(command, null, jarFile.getParentFile()));
        System.err.println(output[0]);
        System.out.println(output[1]);
    }
`

Thanks again.

**Comment by lygoing** [comment]
`package my.examples;

import javassist.ClassPool;
import javassist.CtClass;
import javassist.CtConstructor;
import org.apache.commons.io.FileUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.net.URISyntaxException;
import java.net.URLClassLoader;
import java.security.ProtectionDomain;

/**
 * @author Mr. Luo
 */
public class JavassistExamples {

    private static final Logger LOGGER = LoggerFactory.getLogger(JavassistExamples.class);

    public static void main(String[] args) throws Exception {
        String className = "com.zaxxer.hikari.HikariDataSource";

        CtClass ctClass = ClassPool.getDefault().getCtClass(className);
        CtConstructor ctConstructor = ctClass.makeClassInitializer();
        ctConstructor.setBody("System.out.println(\"Hello Javassist!\");");

        byte[] bytes = ctClass.toBytecode();
        ctClass.detach();

        //Write new class file
        File jarFile = getJarFile(className);
        String newClassPath = jarFile.getParentFile().getAbsolutePath() +
                File.separator + className.replace('.', File.separatorChar) + ".class";
        FileUtils.writeByteArrayToFile(new File(newClassPath), bytes);

        CommandLine commandLine = new CommandLine();

        //Uninstall the application jar package
        URLClassLoader classLoader = (URLClassLoader) JavassistExamples.class.getClassLoader();
        classLoader.close();

        //Write new class file to the jar package
        String[] command = new String[]{"jar", "-uvf", jarFile.getName(), className.replace('.', '/') + ".class"};
        String[] output = commandLine.execute(Runtime.getRuntime().exec(command, null, jarFile.getParentFile()));
        System.err.println(output[0]);
        System.out.println(output[1]);
    }

    public static File getJarFile(String className) throws ClassNotFoundException, URISyntaxException {
        Class clazz = Class.forName(className);
        ProtectionDomain protectionDomain = clazz.getProtectionDomain();
        File jarFile = new File(protectionDomain.getCodeSource().getLocation().toURI());
        LOGGER.debug(jarFile.getParentFile().getAbsolutePath());
        return jarFile;
    }
}
`

**Comment by lygoing** [comment]
```
package my.examples;

import javassist.ClassPool;
import javassist.CtClass;
import javassist.CtConstructor;
import org.apache.commons.io.FileUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.net.URISyntaxException;
import java.net.URLClassLoader;
import java.security.ProtectionDomain;

/**
 * @author Mr. Luo
 */
public class JavassistExamples {

    private static final Logger LOGGER = LoggerFactory.getLogger(JavassistExamples.class);

    public static void main(String[] args) throws Exception {
        String className = "com.zaxxer.hikari.HikariDataSource";

        CtClass ctClass = ClassPool.getDefault().getCtClass(className);
        CtConstructor ctConstructor = ctClass.makeClassInitializer();
        ctConstructor.setBody("System.out.println(\"Hello Javassist!\");");

        byte[] bytes = ctClass.toBytecode();
        ctClass.detach();

        //Write new class file
        File jarFile = getJarFile(className);
        String newClassPath = jarFile.getParentFile().getAbsolutePath() +
                File.separator + className.replace('.', File.separatorChar) + ".class";
        FileUtils.writeByteArrayToFile(new File(newClassPath), bytes);

        CommandLine commandLine = new CommandLine();

        //Uninstall the application jar package
        URLClassLoader classLoader = (URLClassLoader) JavassistExamples.class.getClassLoader();
        classLoader.close();

        //Write new class file to the jar package
        String[] command = new String[]{"jar", "-uvf", jarFile.getName(), className.replace('.', '/') + ".class"};
        String[] output = commandLine.execute(Runtime.getRuntime().exec(command, null, jarFile.getParentFile()));
        System.err.println(output[0]);
        System.out.println(output[1]);
    }

    public static File getJarFile(String className) throws ClassNotFoundException, URISyntaxException {
        Class clazz = Class.forName(className);
        ProtectionDomain protectionDomain = clazz.getProtectionDomain();
        File jarFile = new File(protectionDomain.getCodeSource().getLocation().toURI());
        LOGGER.debug(jarFile.getParentFile().getAbsolutePath());
        return jarFile;
    }
}

```

**Comment by lygoing** [comment]
After I commented on the javassist part of the code, the program works fine. So I think javassist has made a jar package.

**Comment by lygoing** [comment]
我换了一种方式使程序执行成功了。但是核心的问题我还没有弄明白，我Java基础还是太薄弱了，还需要多学习。谢谢您抽空回答我的问题，祝你有个愉快的假期。
```
package my.examples;

import javassist.ClassPool;
import javassist.CtClass;
import javassist.CtConstructor;
import org.apache.commons.io.FileUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.net.URISyntaxException;
import java.net.URLClassLoader;
import java.security.ProtectionDomain;
import java.util.UUID;

/**
 * @author Mr. Luo
 */
public class JavassistExamples {

    private static final Logger LOGGER = LoggerFactory.getLogger(JavassistExamples.class);

    public static void main(String[] args) throws Exception {
        String className = "com.zaxxer.hikari.HikariDataSource";
        Class.forName(className);

        CtClass ctClass = new ClassPool(true).getCtClass(className);
        CtConstructor ctConstructor = ctClass.makeClassInitializer();
        ctConstructor.setBody("System.out.println(\"Hello Javassist!\");");

        byte[] bytes = ctClass.toBytecode();
        ctClass.detach();

        //Write new class file
        File jarFile = getJarFile(className);
        FileUtils.copyFile(jarFile, new File(jarFile.getAbsolutePath() + ".bak"));

        //Create a temporary folder for modifying the write. After the write is successful, replace the original jar file.
        File tmpDir = new File(jarFile.getParentFile() + File.separator + UUID.randomUUID().toString());
        String newClassPath = tmpDir.getAbsolutePath() +
                File.separator + className.replace('.', File.separatorChar) + ".class";
        FileUtils.writeByteArrayToFile(new File(newClassPath), bytes);

        File tmpFile = new File(tmpDir + File.separator + jarFile.getName());
        FileUtils.copyFile(jarFile, tmpFile);

        CommandLine commandLine = new CommandLine();

        //Uninstall the application jar package
        URLClassLoader classLoader = (URLClassLoader) JavassistExamples.class.getClassLoader();
        classLoader.close();

        //Write new class file to the jar package
        String[] command = new String[]{"jar", "-uvf", tmpFile.getName(), className.replace('.', '/') + ".class"};
        String[] output = commandLine.execute(Runtime.getRuntime().exec(command, null, tmpDir));
        System.out.println(output[0]);
        System.err.println(output[1]);

        //Replace the jar package under the classpath with the modified jar package.
        FileUtils.copyFile(tmpFile, jarFile);
        //Delete temporary folder
        FileUtils.deleteDirectory(tmpDir);
    }

    public static File getJarFile(String className) throws ClassNotFoundException, URISyntaxException {
        Class clazz = Class.forName(className);
        ProtectionDomain protectionDomain = clazz.getProtectionDomain();
        File jarFile = new File(protectionDomain.getCodeSource().getLocation().toURI());
        LOGGER.debug(jarFile.getParentFile().getAbsolutePath());
        return jarFile;
    }
}

```

**Comment by lygoing** [comment]
I changed the way to make the program execute successfully. But I still haven't figured out the core problem. My Java foundation is still too weak, and I need to learn more. Thank you for taking the time to answer my questions and wish you a happy holiday.

```
package my.examples;

import javassist.ClassPool;
import javassist.CtClass;
import javassist.CtConstructor;
import org.apache.commons.io.FileUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.net.URISyntaxException;
import java.net.URLClassLoader;
import java.security.ProtectionDomain;
import java.util.UUID;

/**
 * @author Mr. Luo
 */
public class JavassistExamples {

    private static final Logger LOGGER = LoggerFactory.getLogger(JavassistExamples.class);

    public static void main(String[] args) throws Exception {
        String className = "com.zaxxer.hikari.HikariDataSource";
        Class.forName(className);

        CtClass ctClass = new ClassPool(true).getCtClass(className);
        CtConstructor ctConstructor = ctClass.makeClassInitializer();
        ctConstructor.setBody("System.out.println(\"Hello Javassist!\");");

        byte[] bytes = ctClass.toBytecode();
        ctClass.detach();

        //Write new class file
        File jarFile = getJarFile(className);
        FileUtils.copyFile(jarFile, new File(jarFile.getAbsolutePath() + ".bak"));

        //Create a temporary folder for modifying the write. After the write is successful, replace the original jar file.
        File tmpDir = new File(jarFile.getParentFile() + File.separator + UUID.randomUUID().toString());
        String newClassPath = tmpDir.getAbsolutePath() +
                File.separator + className.replace('.', File.separatorChar) + ".class";
        FileUtils.writeByteArrayToFile(new File(newClassPath), bytes);

        File tmpFile = new File(tmpDir + File.separator + jarFile.getName());
        FileUtils.copyFile(jarFile, tmpFile);

        CommandLine commandLine = new CommandLine();

        //Uninstall the application jar package
        URLClassLoader classLoader = (URLClassLoader) JavassistExamples.class.getClassLoader();
        classLoader.close();

        //Write new class file to the jar package
        String[] command = new String[]{"jar", "-uvf", tmpFile.getName(), className.replace('.', '/') + ".class"};
        String[] output = commandLine.execute(Runtime.getRuntime().exec(command, null, tmpDir));
        System.out.println(output[0]);
        System.err.println(output[1]);

        //Replace the jar package under the classpath with the modified jar package.
        FileUtils.copyFile(tmpFile, jarFile);
        //Delete temporary folder
        FileUtils.deleteDirectory(tmpDir);
    }

    public static File getJarFile(String className) throws ClassNotFoundException, URISyntaxException {
        Class clazz = Class.forName(className);
        ProtectionDomain protectionDomain = clazz.getProtectionDomain();
        File jarFile = new File(protectionDomain.getCodeSource().getLocation().toURI());
        LOGGER.debug(jarFile.getParentFile().getAbsolutePath());
        return jarFile;
    }
}

```

**Comment by chibash** [comment]
In the initial example, what error did you observe?  An exception?  Or simply, did you see any update of the class file?
